Semantic Segmentation messages (#71)
Kukanani/vision_msgs_proposal: README.md

in your code, as the message's header should match the header of the source
data.

> Semantic segmentation pipelines should use `sensor_msgs/Image` messages for publishing segmentation and confidence masks. This allows to use all the ROS tools for image processing easily and to choose the most lightweight encoding for each type of message. To transmit the metadata associated with the vision pipeline the [`/vision_msgs/InferenceInfo`](msg/InferenceInfo.msg) message can be used analogously to how `/sensor_msgs/CameraInfo` message and `/sensor_msgs/Image` are:

Reviewer: ```suggestion
Semantic segmentation pipelines should use `sensor_msgs/Image` messages for publishing segmentation and confidence masks. This allows systems to use standard ROS tools for image processing, and allows choosing the most compact image encoding appropriate for the task. To transmit the metadata associated with the vision pipeline, you should use the [`/vision_msgs/InferenceInfo`](msg/InferenceInfo.msg) message. This message works the same as `/sensor_msgs/CameraInfo or [`/vision_msgs/VisionInfo`](msg/VisionInfo.msg):
```

> 1. The `InferenceInfo` topic can be latched so that new nodes joining the ROS system can get the messages that were published since the beginning. In ROS2 this can be achieved using a `transient local` QoS profile.

Reviewer 2: It should be `inference_info` when talking about the topic  (as opposed to the message type `InferenceInfo`).

> 1. The `InferenceInfo` topic can be latched so that new nodes joining the ROS system can get the messages that were published since the beginning. In ROS2 this can be achieved using a `transient local` QoS profile.

Reviewer: ```suggestion
1. Publish `InferenceInfo` to a topic. The topic should be at same namespace level as the associated image. That is, if your image is published at `/my_segmentation_node/image`, the `InferenceInfo` should be published at `/my_segmentation_node/inference_info`. Use a latched publisher for `InferenceInfo`, so that new nodes joining the ROS system can get the messages that were published since the beginning. In ROS2, this can be achieved using a `transient local` QoS profile.
```

> 2. The subscribing node can get and store one `InferenceInfo` message and cancel its subscription after that. This assumes the provider of the message publishes it periodically.

Reviewer: ```suggestion
2. A subscribing node may receive an `InferenceInfo` message, store it locally, then cancel its subscription.
```

Reviewer: I don't understand why we need to assume that the publisher publishes periodically - that's the point of making it latched

Author: Its my understanding that a latched topic uses the transient local QoS. In contrast, this second method does not require the usage of any specific QoS, that's why messages should be published periodically to allow new nodes to get the messages.

Reviewer: Makes sense

> 3. The `InferenceInfo` message can be synchronized with the `Image` messages that containing the segmentation mask and the confidence values of the inference of each pixel. This can be achieved using ROS's `message_filters`. The same [snippet](http://wiki.ros.org/message_filters#Policy-Based_Synchronizer_.5BROS_1.1.2B-.5D) mentioned above can work as an example.

Reviewer 2: Isn't this incompatible with mentioning the latched topic above? If we use the time synchronizers from `message_filters`, we need to publish a `InferenceInfo` message for every image with the exact same time stamp (just like the `camera_info` topic). The section about latched publishers above sounds like we only publish a message once (or whenever the contents change), analogous to static TFs or maps.

I'm fine with both, BTW.

Reviewer: Agreed. `InferenceInfo` should be one of these two:
1. A message truly like `CameraInfo` - it is not expected to change. Caching it locally on other nodes should be fine. Latched publishers should be fine.
2. It can't be assumed to be the same for each published `Image`. If that's the case, we need to remove all mention of latching and caching, and require that people subscribe to both the `Image` and the `InferenceInfo`. But if they're always being published in pairs, you'll need to remind me why we aren't combining them into one message.

Author: I think it can be both ways, that's why I want to place both options in the readme to describe what would be the common ways of using the message. The third option I propose also assumes the message does not change but does not require the use of a latched topic. I think that it is worth referencing at least the first two but let me know if you think we should settle for only one of them
